File: packages/core/ios/Sources/DdRumImplementation.swift
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

import Foundation
@_spi(Experimental)
import DatadogRUM
import DatadogInternal
import React

private extension RUMActionType {
    init(from string: String) {
        switch string.lowercased() {
        case "tap": self = .tap
        case "scroll": self = .scroll
        case "swipe": self = .swipe
        default: self = .custom
        }
    }
}

internal extension RUMErrorSource {
    init(from string: String) {
        switch string.lowercased() {
        case "source": self = .source
        case "network": self = .network
        case "webview": self = .webview
        case "console": self = .console
        default: self = .custom
        }
    }
}

private extension RUMResourceType {
    init(from string: String) {
        switch string {
        case "image": self = .image
        case "xhr": self = .xhr
        case "beacon": self = .beacon
        case "css": self = .css
        case "document": self = .document
        case "fetch": self = .fetch
        case "font": self = .font
        case "js": self = .js
        case "media": self = .media
        default: self = .other
        }
    }
}

private extension RUMMethod {
    init(from string: String) {
        switch string.uppercased() {
        case "POST": self = .post
        case "GET": self = .get
        case "HEAD": self = .head
        case "PUT": self = .put
        case "DELETE": self = .delete
        case "PATCH": self = .patch
        default: self = .get
        }
    }
}

internal extension RUMFeatureOperationFailureReason {
    init(from string: String) {
        switch string.lowercased() {
        case "error": self = .error
        case "abandoned": self = .abandoned
        default: self = .other
        }
    }
}

@objc
public class DdRumImplementation: NSObject {
    internal static let timestampKey = "_dd.timestamp"
    internal static let fingerprintKey = "_dd.error.fingerprint"
    internal static let resourceTimingsKey = "_dd.resource_timings"

    internal static let fetchTimingKey = "fetch"
    internal static let redirectTimingKey = "redirect"
    internal static let dnsTimingKey = "dns"
    internal static let connectTimingKey = "connect"
    internal static let sslTimingKey = "ssl"
    internal static let firstByteTimingKey = "firstByte"
    internal static let downloadTimingKey = "download"

    internal static let missingResourceSize = -1

    lazy var nativeRUM: RUMMonitorProtocol = rumProvider()
    lazy var rumInternal: RUMMonitorInternalProtocol? = rumInternalProvider()
    lazy var heatmapIdentifierRegistry: HeatmapIdentifierRegistry? = heatmapIdentifierRegistryProvider()
    private let mainDispatchQueue: DispatchQueueType
    private let uiManager: RCTUIManager
    private let rootViewProvider: () -> UIView?
    private let heatmapIdentifierRegistryProvider: () -> HeatmapIdentifierRegistry?
    private let rumProvider: () -> RUMMonitorProtocol
    private let rumInternalProvider: () -> RUMMonitorInternalProtocol?

    private typealias UserAction = (type: RUMActionType, name: String?)

    internal init(
        mainDispatchQueue: DispatchQueueType,
        uiManager: RCTUIManager,
        rootViewProvider: @escaping () -> UIView?,
        heatmapIdentifierRegistryProvider: @escaping () -> HeatmapIdentifierRegistry?,
        rumProvider: @escaping () -> RUMMonitorProtocol,
        rumInternalProvider: @escaping () -> RUMMonitorInternalProtocol?
    ) {
        self.mainDispatchQueue = mainDispatchQueue
        self.uiManager = uiManager
        self.rootViewProvider = rootViewProvider
        self.heatmapIdentifierRegistryProvider = heatmapIdentifierRegistryProvider
        self.rumProvider = rumProvider
        self.rumInternalProvider = rumInternalProvider
    }

    @objc
    public convenience init(bridge: RCTBridge) {
        self.init(
            mainDispatchQueue: DispatchQueue.main,
            uiManager: bridge.uiManager,
            rootViewProvider: { UIWindow.reactRootView() },
            heatmapIdentifierRegistryProvider: { CoreRegistry.default.heatmapIdentifierRegistry },
            rumProvider: { RUMMonitor.shared() },
            rumInternalProvider: { RUMMonitor.shared()._internal }
        )
    }

    @objc
    public func startView(key: String, name: String, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.startView(key: key, name: name, attributes: attributes(from: context, with: timestampMs))
        resolve(nil)
    }

    @objc
    public func stopView(key: String, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.stopView(key: key, attributes: attributes(from: context, with: timestampMs))
        resolve(nil)
    }

    @objc
    public func startAction(type: String, name: String, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.startAction(type: RUMActionType(from: type), name: name, attributes: attributes(from: context, with: timestampMs))
        resolve(nil)
    }

    @objc
    public func stopAction(type: String, name: String, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.stopAction(type: RUMActionType(from: type), name: name, attributes: attributes(from: context, with: timestampMs))
        resolve(nil)
    }

    @objc
    public func addAction(type: String, name: String, touch: NSDictionary?, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        if
            let touch,
            let reactTag = touch["reactTag"] as? NSNumber,
            let x = touch["x"] as? NSNumber,
            let y = touch["y"] as? NSNumber,
            let pageX = touch["pageX"] as? NSNumber,
            let pageY = touch["pageY"] as? NSNumber
        {
            addAction(
                at: Date(timeIntervalSince1970: timestampMs / 1_000),
                type: RUMActionType(from: type),
                name: name,
                reactTag: reactTag,
                location: .init(x: CGFloat(truncating: x), y: CGFloat(truncating: y)),
                pageLocation: .init(x: CGFloat(truncating: pageX), y: CGFloat(truncating: pageY)),
                attributes: castAttributesToSwift(context)
            )
        } else {
            nativeRUM.addAction(type: RUMActionType(from: type), name: name, attributes: attributes(from: context, with: timestampMs))
        }
        resolve(nil)
    }

    @objc
    public func startResource(key: String, method: String, url: String, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.startResource(resourceKey: key, httpMethod: RUMMethod(from: method), urlString: url, attributes: attributes(from: context, with: timestampMs))
        resolve(nil)
    }

    @objc
    public func stopResource(key: String, statusCode: Int64, kind: String, size: Double, context: NSDictionary, timestampMs: Double, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        let mutableContext = NSMutableDictionary(dictionary: context)
        if let resourceTimings = mutableContext.object(forKey: Self.resourceTimingsKey) as? [String: Any] {
            mutableContext.removeObject(forKey: Self.resourceTimingsKey)

            addResourceMetrics(key: key, resourceTimings: resourceTimings)
        }

        nativeRUM.stopResource(
            resourceKey: key,
            statusCode: Int(statusCode),
            kind: RUMResourceType(from: kind),
            size: Int64(size) == Self.missingResourceSize ? nil : Int64(size),
            attributes: attributes(from: mutableContext, with: timestampMs)
        )
        resolve(nil)
    }

    @objc
    public func addError(message: String, source: String, stacktrace: String, context: NSDictionary, timestampMs: Double, fingerprint: String, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
    
        func addErrorWithContext(errorContext: NSDictionary) -> Void {
            nativeRUM.addError(message: message, type: nil, stack: stacktrace, source: RUMErrorSource(from: source), attributes: attributes(from: errorContext, with: timestampMs), file: nil, line: nil)
        }
        
        if !fingerprint.isEmpty {
            let updatedContext = NSMutableDictionary(dictionary: context)
            updatedContext[Self.fingerprintKey] = fingerprint
            addErrorWithContext(errorContext: updatedContext)
        } else {
            addErrorWithContext(errorContext: context)
        }
        
        resolve(nil)
    }

    @objc
    public func addTiming(name: String, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.addTiming(name: name)
        resolve(nil)
    }
    
    @objc
    public func addViewAttribute(key: AttributeKey, value: NSDictionary, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        if let attributeValue = value.object(forKey: "value") {
            let castedAttribute = castValueToSwift(attributeValue)
            nativeRUM.addViewAttribute(forKey: key, value: castedAttribute)
        }
        resolve(nil)
    }
    
    @objc
    public func removeViewAttribute(key: AttributeKey, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.removeViewAttribute(forKey: key)
        resolve(nil)
    }
    
    @objc
    public func addViewAttributes(attributes: NSDictionary, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        let castedAttributes = castAttributesToSwift(attributes)
        nativeRUM.addViewAttributes(castedAttributes)
        resolve(nil)
    }
    
    @objc
    public func removeViewAttributes(keys: [AttributeKey], resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.removeViewAttributes(forKeys: keys)
        resolve(nil)
    }
    
    @objc
    public func addViewLoadingTime(overwrite: Bool, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.addViewLoadingTime(overwrite: overwrite)
        resolve(nil)
    }

    @objc
    public func reportAppFullyDisplayed(resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.reportAppFullyDisplayed()
        resolve(nil)
    }

    @objc
    public func stopSession(resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        nativeRUM.stopSession()
        resolve(nil)
    }

    @objc
    public func addFeatureFlagEvaluation(name: String, value: NSDictionary, resolve:RCTPromiseResolveBlock, reject:RCTPromiseRejectBlock) -> Void {
        let valueAsEncodable = castAttributesToSwift(value)
        if let value = valueAsEncodable["value"] {
            nativeRUM.addFeatureFlagEvaluation(name: name, value: value)
        }
        resolve(nil)
    }
    
    @objc
    public func getCurrentSessionId(_ resolve: @escaping (Any?) -> Void, reject: RCTPromiseRejectBlock) -> Void {
        nativeRUM.currentSessionID { sessionId in
            resolve(sessionId)
        }
    }
    
    @objc
    public func startFeatureOperation(
        name: String,
        operationKey: String?,
        attributes: NSDictionary,
        resolve: @escaping (Any?) -> Void,
        reject: RCTPromiseRejectBlock
    ){
        let castedAttributes = castAttributesToSwift(attributes)
        nativeRUM.startFeatureOperation(name: name, operationKey: operationKey, attributes: castedAttributes)
        resolve(nil)
    }

    @objc
    public func succeedFeatureOperation(
        name: String,
        operationKey: String?,
        attributes: NSDictionary,
        resolve: @escaping (Any?) -> Void,
        reject: RCTPromiseRejectBlock
    ){
        let castedAttributes = castAttributesToSwift(attributes)
        nativeRUM.succeedFeatureOperation(name: name, operationKey: operationKey, attributes: castedAttributes)
        resolve(nil)
    }

    @objc
    public func failFeatureOperation(
        name: String,
        operationKey: String?,
        reason: String,
        attributes: NSDictionary,
        resolve: @escaping (Any?) -> Void,
        reject: RCTPromiseRejectBlock
    ){
        let castedAttributes = castAttributesToSwift(attributes)
        nativeRUM.failFeatureOperation(name: name, operationKey: operationKey,
                                       reason: RUMFeatureOperationFailureReason(from: reason), attributes: castedAttributes)
        resolve(nil)
    }

    // MARK: - Private methods

    private func attributes(from context: NSDictionary, with timestampMs: Double) -> [String: Encodable] {
        var context = context as? [String: Any] ?? [:]
        context[Self.timestampKey] = Int64(timestampMs)
        return castAttributesToSwift(context)
    }

    private func addResourceMetrics(key: String, resourceTimings: [String: Any]) {
        let fetch = timingValue(from: resourceTimings, for: Self.fetchTimingKey)
        let redirect = timingValue(from: resourceTimings, for: Self.redirectTimingKey)
        let dns = timingValue(from: resourceTimings, for: Self.dnsTimingKey)
        let connect = timingValue(from: resourceTimings, for: Self.connectTimingKey)
        let ssl = timingValue(from: resourceTimings, for: Self.sslTimingKey)
        let firstByte = timingValue(from: resourceTimings, for: Self.firstByteTimingKey)
        let download = timingValue(from: resourceTimings, for: Self.downloadTimingKey)

        
        if let fetch = fetch {
            rumInternal?.addResourceMetrics(
                at: Date.init(),
                resourceKey: key,
                fetch: fetch,
                redirection: redirect,
                dns: dns,
                connect: connect,
                ssl: ssl,
                firstByte: firstByte,
                download: download,
                // no need to define the size here, because if it is missing,
                // it will be taken from the command
                responseBodySize: nil,
                requestBodySize: nil,
                attributes: [:]
            )
        }
    }

    private func addAction(
        at time: Date,
        type: RUMActionType,
        name: String,
        reactTag: NSNumber,
        location: CGPoint,
        pageLocation: CGPoint,
        attributes: [AttributeKey: AttributeValue]
    ) {
        mainDispatchQueue.async { [uiManager, rootViewProvider, heatmapIdentifierRegistry, rumInternal] in
            var location = location
            let view = uiManager.view(
                forReactTag: reactTag,
                location: &location,
                pageLocation: pageLocation,
                rootViewProvider: rootViewProvider
            )

            let heatmapAttributes: HeatmapAttributes? = view.flatMap { view in
                guard let identifier = heatmapIdentifierRegistry?.heatmapIdentifier(for: ObjectIdentifier(view)) else {
                    return nil
                }
                return HeatmapAttributes(
                    identifier: identifier,
                    size: view.bounds.size,
                    location: location
                )
            }
            rumInternal?.addAction(
                at: time,
                type: type,
                name: name,
                heatmapAttributes: heatmapAttributes,
                attributes: attributes
            )
        }
    }

    private func timingValue(from timings: [String: Any], for timingName: String) -> (start: Date, end: Date)? {
        let timing = timings[timingName] as? [String: NSNumber]
        if let startInNs = timing?["startTime"]?.int64Value, let durationInNs = timing?["duration"]?.int64Value {
            return (
                Date(timeIntervalSince1970: TimeInterval(fromNs: startInNs)),
                Date(timeIntervalSince1970: TimeInterval(fromNs: startInNs + durationInNs))
            )
        }
        return nil
    }
}

internal extension TimeInterval {
    init(fromNs ns: Int64) { self = TimeInterval(Double(ns) / 1_000_000_000) }
}

File: packages/core/android/src/main/kotlin/com/datadog/reactnative/DdRumImplementation.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2016-Present Datadog, Inc.
 */

package com.datadog.reactnative

import android.util.Log
import com.datadog.android.rum.RumActionType
import com.datadog.android.rum.RumAttributes
import com.datadog.android.rum.RumErrorSource
import com.datadog.android.rum.RumResourceKind
import com.datadog.android.rum.RumResourceMethod
import com.datadog.android.rum.featureoperations.FailureReason
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import java.util.Locale

/**
 * The entry point to use Datadog's RUM feature.
 */
@Suppress("TooManyFunctions")
class DdRumImplementation internal constructor(
    private val datadog: DatadogWrapper = DatadogSDKWrapper(),
    private val heatmapActionHandler: HeatmapActionHandler = HeatmapActionHandler()
) {
    /**
     * Start tracking a RUM View.
     * @param key The view unique key identifier.
     * @param name The view name.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the view started (in milliseconds). If not provided, current timestamp will be used.
     */
    fun startView(
        key: String,
        name: String,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }
        datadog.getRumMonitor().startView(
            key = key,
            name = name,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Stop tracking a RUM View.
     * @param key The view unique key identifier.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the view stopped (in milliseconds). If not provided, current timestamp will be used.
     */
    fun stopView(key: String, context: ReadableMap, timestampMs: Double, promise: Promise) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }
        datadog.getRumMonitor().stopView(
            key = key,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Start tracking a RUM Action.
     * @param type The action type (tap, scroll, swipe, click, custom).
     * @param name The action name.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the action started (in milliseconds). If not provided, current timestamp will be used.
     */
    fun startAction(
        type: String,
        name: String,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }
        datadog.getRumMonitor().startAction(
            type = type.asRumActionType(),
            name = name,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Stop tracking the ongoing RUM Action.
     * @param type The action type (tap, scroll, swipe, click, custom).
     * @param name The action name.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the action stopped (in milliseconds). If not provided, current timestamp will be used.
     */
    fun stopAction(
        type: String,
        name: String,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }
        datadog.getRumMonitor().stopAction(
            type = type.asRumActionType(),
            name = name,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Add a RUM Action.
     * @param type The action type (tap, scroll, swipe, click, custom).
     * @param name The action name.
     * @param touch The native touch data for tap actions, or null for other action types.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the action occurred (in milliseconds). If not provided, current timestamp will be used.
     */
    @Suppress("LongParameterList")
    fun addAction(
        type: String,
        name: String,
        touch: ReadableMap?,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }

        val eligibleAction = heatmapActionHandler.resolveEligibility(datadog, type, name, touch)
        if (eligibleAction == null) {
            addActionWithoutHeatmap(type, name, attributes)
            promise.resolve(null)
            return
        }

        // Resolve before dispatching heatmap work (iOS parity).
        promise.resolve(null)

        heatmapActionHandler.attachHeatmapData(eligibleAction, name, attributes) {
            addActionWithoutHeatmap(type, name, attributes)
        }
    }

    private fun addActionWithoutHeatmap(type: String, name: String, attributes: Map<String, Any?>) {
        datadog.getRumMonitor().addAction(
            type = type.asRumActionType(),
            name = name,
            attributes = attributes
        )
    }

    /**
     * Start tracking a RUM Resource.
     * @param key The resource unique key identifier.
     * @param method The resource method (GET, POST, …).
     * @param url The resource url.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the resource started (in milliseconds). If not provided, current timestamp will be used.
     */
    @Suppress("LongParameterList")
    fun startResource(
        key: String,
        method: String,
        url: String,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }
        datadog.getRumMonitor().startResource(
            key = key,
            method = method.asRumResourceMethod(),
            url = url,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Stop tracking a RUM Resource.
     * @param key The resource unique key identifier.
     * @param statusCode The resource status code.
     * @param kind The resource's kind (xhr, document, image, css, font, …).
     * @param size The resource size in bytes.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the resource stopped (in milliseconds). If not provided, current timestamp will be used.
     */
    @Suppress("LongParameterList")
    fun stopResource(
        key: String,
        statusCode: Double,
        kind: String,
        size: Double,
        context: ReadableMap,
        timestampMs: Double,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }

        val resourceSize = if (size.toLong() == MISSING_RESOURCE_SIZE) {
            null
        } else {
            size.toLong()
        }
        datadog.getRumMonitor().stopResource(
            key = key,
            statusCode = statusCode.toInt(),
            kind = kind.asRumResourceKind(),
            size = resourceSize,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Add a RUM Error.
     * @param message The error message.
     * @param source The error source (network, source, console, logger, …).
     * @param stacktrace The error stacktrace.
     * @param context The additional context to send.
     * @param timestampMs The timestamp when the error occurred (in milliseconds). If not provided, current timestamp will be used.
     * @param fingerprint A custom fingerprint to group this error with similar ones.
     */
    @Suppress("LongParameterList")
    fun addError(
        message: String,
        source: String,
        stacktrace: String,
        context: ReadableMap,
        timestampMs: Double,
        fingerprint: String,
        promise: Promise
    ) {
        val attributes = context.toHashMap().toMutableMap().apply {
            put(RumAttributes.INTERNAL_TIMESTAMP, timestampMs.toLong())
        }

        if (fingerprint.isNotEmpty()) {
            attributes[RumAttributes.ERROR_FINGERPRINT] = fingerprint
        }

        datadog.getRumMonitor().addErrorWithStacktrace(
            message = message,
            source = source.asErrorSource(),
            stacktrace = stacktrace,
            attributes = attributes
        )
        promise.resolve(null)
    }

    /**
     * Adds a specific timing in the active View. The timing duration will be computed as the difference between the time the View was started and the time this function was called.
     * @param name The name of the new custom timing attribute. Timings can be nested up to 8 levels deep. Names using more than 8 levels will be sanitized by SDK.
     */
    fun addTiming(name: String, promise: Promise) {
        datadog.getRumMonitor().addTiming(name)
        promise.resolve(null)
    }

    /**
     * Adds a custom attribute to the active RUM View. It will be propagated to all future RUM events associated with the active View.
     * @param key: key for this view attribute.
     * @param value: value for this attribute.
     */
    fun addViewAttribute(key: String, value: ReadableMap, promise: Promise) {
        val attributeValue = value.toMap()["value"]
        val attributes = mutableMapOf<String, Any?>()
        attributes[key] = attributeValue
        datadog.getRumMonitor().addViewAttributes(attributes)
        promise.resolve(null)
    }

    /**
     * Removes an attribute from the active RUM View.
     * @param key: key for the attribute to be removed from the view.
     */
    fun removeViewAttribute(key: String, promise: Promise) {
        val keysToDelete: Collection<String> = listOf(key)
        datadog.getRumMonitor().removeViewAttributes(keysToDelete)
        promise.resolve(null)
    }

    /**
     * Adds multiple attributes to the active RUM View. They will be propagated to all future RUM events associated with the active View.
     * @param attributes: key/value object containing all attributes to be added to the view.
     */
    fun addViewAttributes(attributes: ReadableMap, promise: Promise) {
        datadog.getRumMonitor().addViewAttributes(attributes.toMap())
        promise.resolve(null)
    }

    /**
     * Removes multiple attributes from the active RUM View.
     * @param keys: keys for the attributes to be removed from the view.
     */
    fun removeViewAttributes(keys: ReadableArray, promise: Promise) {
        val keysToDelete = (0 until keys.size())
            .mapNotNull { keys.getString(it) }
        datadog.getRumMonitor().removeViewAttributes(keysToDelete)
        promise.resolve(null)
    }

    /**
     * Adds the loading time of the view to the active view.
     * It is calculated as the difference between the current time and the start time of the view.
     * @param overwrite: If true, overwrites the previously calculated view loading time.
     */
    fun addViewLoadingTime(overwrite: Boolean, promise: Promise) {
        datadog.getRumMonitor().addViewLoadingTime(overwrite)
        promise.resolve(null)
    }

    /**
     * This method can be used to mark the moment in time when the UI of the app is considered fully displayed.
     * The duration between the application launch and this moment of time will be shown as TTFD (time to full display)
     * in the RUM session explorer. Only the first call to this method will have any effect for a given RUM session.
     */
    fun reportAppFullyDisplayed(promise: Promise) {
        datadog.getRumMonitor().reportAppFullyDisplayed()
        promise.resolve(null)
    }

    /**
     * Stops the current RUM Session.
     */
    fun stopSession(promise: Promise) {
        datadog.getRumMonitor().stopSession()
        promise.resolve(null)
    }

    /**
     * Adds result of evaluating a feature flag to the view.
     * Feature flag evaluations are local to the active view and are cleared when the view is stopped.
     * @param name The name of the feature flag
     * @param valueAsMap The value the feature flag evaluated to, encapsulated in a Map
     */
    fun addFeatureFlagEvaluation(name: String, valueAsMap: ReadableMap, promise: Promise) {
        val value = valueAsMap.toHashMap()["value"]
        if (value != null) {
            datadog.getRumMonitor().addFeatureFlagEvaluation(name, value)
        }
        promise.resolve(null)
    }

    /**
     * Returns current session ID, or null if unavailable.
     */
    fun getCurrentSessionId(promise: Promise) {
        datadog.getRumMonitor().getCurrentSessionId {
            promise.resolve(it)
        }
    }

    /**
     * Starts a Feature Operation.
     *
     * @param name Human-readable operation name (e.g., "login_flow").
     * @param operationKey Optional key that uniquely identifies this operation instance.
     * @param attributes Additional attributes to attach to the operation.
     * @param promise Resolved with `null` when the call completes.
     */
    fun startFeatureOperation(name: String, operationKey: String? = null, attributes: ReadableMap, promise: Promise) {
        val attributesMap = attributes.toHashMap().toMutableMap()
        datadog.getRumMonitor().startFeatureOperation(name, operationKey, attributesMap);
        promise.resolve(null)
    }

    /**
     * Marks a Feature Operation as successfully completed.
     *
     * @param name The name of the feature operation (for example, `"login_flow"`).
     * @param operationKey The key of the operation instance to complete, if one was provided when starting it.
     * @param attributes A map of custom attributes to attach to this completion event.
     */
    fun succeedFeatureOperation(name: String, operationKey: String? = null, attributes: ReadableMap, promise: Promise) {
        val attributesMap = attributes.toHashMap().toMutableMap()
        datadog.getRumMonitor().succeedFeatureOperation(name, operationKey, attributesMap)
        promise.resolve(null)
    }


    /**
     * Marks a Feature Operation as failed.
     *
     * @param name The name of the feature operation (for example, `"login_flow"`).
     * @param operationKey The key of the operation instance to fail, if one was provided when starting it.
     * @param failureReason The reason for the failure. Possible values are defined in [FailureReason]
     *                      (e.g., `FailureReason.ERROR`, `FailureReason.ABANDONED`, `FailureReason.OTHER`).
     * @param attributes A map of custom attributes to attach to this failure event.
     */
    fun failFeatureOperation(
        name: String,
        operationKey: String? = null,
        failureReason: String,
        attributes: ReadableMap,
        promise: Promise
    ) {
        val attributesMap = attributes.toHashMap().toMutableMap()
        val reason = runCatching {
            enumValueOf<FailureReason>(failureReason.uppercase())
        }.getOrDefault(FailureReason.OTHER)

        datadog.getRumMonitor().failFeatureOperation(name, operationKey, reason, attributesMap)
        promise.resolve(null)
    }

    // region Internal

    private fun String.asRumActionType(): RumActionType {
        return when (lowercase(Locale.US)) {
            "tap" -> RumActionType.TAP
            "scroll" -> RumActionType.SCROLL
            "swipe" -> RumActionType.SWIPE
            "click" -> RumActionType.CLICK
            "back" -> RumActionType.BACK
            else -> RumActionType.CUSTOM
        }
    }

    private fun String.asRumResourceKind(): RumResourceKind {
        return when (lowercase(Locale.US)) {
            "xhr" -> RumResourceKind.XHR
            "native" -> RumResourceKind.NATIVE
            "fetch" -> RumResourceKind.FETCH
            "document" -> RumResourceKind.DOCUMENT
            "beacon" -> RumResourceKind.BEACON
            "js" -> RumResourceKind.JS
            "image" -> RumResourceKind.IMAGE
            "font" -> RumResourceKind.FONT
            "css" -> RumResourceKind.CSS
            "media" -> RumResourceKind.MEDIA
            "other" -> RumResourceKind.OTHER
            else -> RumResourceKind.UNKNOWN
        }
    }

    private fun String.asErrorSource(): RumErrorSource {
        return when (lowercase(Locale.US)) {
            "agent" -> RumErrorSource.AGENT
            "console" -> RumErrorSource.CONSOLE
            "logger" -> RumErrorSource.LOGGER
            "network" -> RumErrorSource.NETWORK
            "source" -> RumErrorSource.SOURCE
            "webview" -> RumErrorSource.WEBVIEW
            "custom" -> RumErrorSource.CUSTOM
            "report" -> RumErrorSource.REPORT
            else -> RumErrorSource.SOURCE
        }
    }

    private fun String.asRumResourceMethod(): RumResourceMethod {
        return when(lowercase(Locale.US)) {
            "get" -> RumResourceMethod.GET
            "delete" -> RumResourceMethod.DELETE
            "head" -> RumResourceMethod.HEAD
            "patch" -> RumResourceMethod.PATCH
            "put" -> RumResourceMethod.PUT
            "post" -> RumResourceMethod.POST
            "trace" -> RumResourceMethod.TRACE
            "options" -> RumResourceMethod.OPTIONS
            "connect" -> RumResourceMethod.CONNECT
            else -> {
                Log.w(
                    DdRum::class.java.canonicalName,
                    "Unknown RUM resource method given: $this, " +
                            "using ${RumResourceMethod.GET} as default"
                )
                RumResourceMethod.GET
            }
        }
    }

    // endregion

    @Suppress("UndocumentedPublicClass")
    companion object {
        private const val MISSING_RESOURCE_SIZE = -1L
        internal const val NAME = "DdRum"
    }
}
